### DISCUSSION TITLE
Propose to fully switch to DDS XTypes

### DISCUSSION THREAD

**Discussion by sonndinh** [OpenDDS/opendds-monitor#44] [General]
The original implementation of the monitor supports a custom way to communicate the topic type that existed before OpenDDS had XTypes dynamic API. The topic type information is encoded in the `topic_data` field of the `PublicationBuiltinTopicData` and `SubscriptionBuiltinTopicData` samples. It has a specific format with:
- a header, and
- a payload for the CDR serialized data of the CORBA `TypeCode` for the corresponding type (see `TopicInfo::storeUserData` for details).

Thus, it only works with applications that use the same format for the data in the `topic_data` field of the publication and subscription built-in topic samples. These applications also have to opt for using CORBA `TypeCode` to convey the type information. For regular OpenDDS users, that would be a rare use case.

Along with the `TypeCode`, a class called `OpenDynamicData` is used to represent data samples of any type. This can be replaced by XTypes DynamicData.

Since OpenDDS has implemented XTypes dynamic API, it makes sense to fully switch to using the XTypes features and remove the use of CORBA `TypeCode` and `OpenDynamicData`. This would leverage the standardized XTypes features and remove the need for the original approach where DDS XTypes has already provided a solution. This would also reduce maintenance efforts for the original approach.

A non-comprehensive list of places that will need updates:
- `OpenDynamicData` (open_dynamic_data.h/.cpp): replaced by XTypes `DynamicData`
- `TypeInfo` and `CommonData` (dds_data.h/.cpp): remove the use of `CORBA::TypeCode` and `CommonData::m_samples` for storing `OpenDynamicData` instances
- `PublicationMonitor` (publication_monitor.h/.cpp) and `SubscriptionMonitor` (subscription_monitor.h/.cpp): remove the use of `topic_data` field in the publication and subscription built-in topic data.
- `TopicMonitor` (topic_monitor.h/.cpp): remove the use of `TopicInfo::typeCode`. Could also remove the use of OpenDDS's `Recorder`.

Probably, there are other places that need to be updated.

**Comment by jrw972** [comment]
@doug1234 What are you thoughts?

**Reply by doug1234** [reply]
It should be done eventually.  Is it possible to have a version that is compatible with both and eventually move to only supporting x types?

**Reply by sonndinh** [reply]
From what I've seen so far, I think it's possible. That sounds like a good transitional version to have.

**Comment by pyphais** [comment]
No idea if my input as a user is useful here or not, but as someone who has spent the last two days trying to add the ability to publish dynamic data samples (and still can't seem to get it to serialize properly when publishing), full xtypes transition would be very appreciated
